Functionbeat 7.9.0 deploy failure on GCP

Deploying functionbeat 7.9.0 fails after the following command:

`./functionbeat -v -e -d "*" deploy storage`

Error output:

~~~"DEBUG   [gcp]   gcp/cli_manager.go:47   Deploy finish for function 'storage'
Function: storage, could not deploy, error: error while creating function (code: 3):
Build failed: src/storage/storage.go:11:2: cannot find package "github.com/elastic/beats/v7/libbeat/beat" in any of:
        /usr/local/go/src/github.com/elastic/beats/v7/libbeat/beat (from $GOROOT)
        /workspace/src/github.com/elastic/beats/v7/libbeat/beat (from $GOPATH)"
~~~
Many additional lines similar to above and finishes with...

~~~"DEBUG   [cli-handler]   cmd/cli_handler.go:62   Deploy execution ended
Fail to deploy 1 function(s)"
~~~

For confirmed bugs, please report:
- Version: functionbeat-7.9.0-linux-x86_64
- Operating System: CentOS-6-1
- Discuss Forum URL: spoke with Noémi Ványi through slack #beats
- Steps to Reproduce: I can see same results with MacOS laptop and GCP provisioned CentOS VM. Previous 7.8.1 version using the same **credentials file** and **functionbeat.yml** file succeeded with following output:

~~~"DEBUG   [gcp]   gcp/op_wait_for_function.go:57  Successfully deployed function
2020-08-27T14:43:11.850Z        DEBUG   [gcp.executor]  executor/executor.go:68 All operations successful
2020-08-27T14:43:11.850Z        DEBUG   [gcp]   gcp/cli_manager.go:50   Successfully created function 'storage'
2020-08-27T14:43:11.850Z        DEBUG   [gcp]   gcp/cli_manager.go:51   Deploy finish for function 'storage'
Function: storage, deploy successful
2020-08-27T14:43:11.850Z        DEBUG   [cli-handler]   cmd/cli_handler.go:64   Deploy execution ended"  
~~~

--- a/CHANGELOG.next.asciidoc
+++ b/CHANGELOG.next.asciidoc
@@ -404,6 +404,7 @@ field. You can revert this change by configuring tags for the module and omittin
 
 - Fix timeout option of GCP functions. {issue}16282[16282] {pull}16287[16287]
 - Do not need Google credentials if not required for the operation. {issue}17329[17329] {pull}21072[21072]
+- Fix dependency issues of GCP functions. {issue}20830[20830] {pull}21070[21070]
 
 ==== Added
 

--- a/dev-tools/mage/gotool/go.go
+++ b/dev-tools/mage/gotool/go.go
@@ -85,6 +85,25 @@ func ListDeps(pkg string) ([]string, error) {
 	return getLines(callGo(nil, "list", "-deps", "-f", tmpl, pkg))
 }
 
+// ListDepsLocation calls `go list -dep` for every package spec given.
+func ListDepsLocation(pkg string) (map[string]string, error) {
+	const tmpl = `{{if not .Standard}}{{.ImportPath}};{{.Dir}}{{end}}`
+
+	lines, err := getLines(callGo(nil, "list", "-deps", "-f", tmpl, pkg))
+	if err != nil {
+		return nil, err
+	}
+	deps := make(map[string]string, len(lines))
+	for _, l := range lines {
+		parts := strings.Split(l, ";")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid number of parts")
+		}
+		deps[parts[0]] = parts[1]
+	}
+	return deps, nil
+}
+
 // ListTestFiles lists all go and cgo test files available in a package.
 func ListTestFiles(pkg string) ([]string, error) {
 	const tmpl = `{{ range .TestGoFiles }}{{ printf "%s\n" . }}{{ end }}` +

--- a/x-pack/functionbeat/dev-tools/packaging/packages.yml
+++ b/x-pack/functionbeat/dev-tools/packaging/packages.yml
@@ -65,6 +65,12 @@ shared:
       pkg/functionbeat-aws:
         source: 'provider/aws/build/golang-crossbuild/aws-linux-amd64'
         mode: 0755
+      pkg/pubsub/vendor:
+        source: 'provider/gcp/build/pubsub/vendor'
+        mode: 0644
+      pkg/storage/vendor:
+        source: 'provider/gcp/build/storage/vendor'
+        mode: 0644
       pkg/pubsub/pubsub.go:
         source: 'provider/gcp/pubsub/pubsub.go'
         mode: 0655

--- a/x-pack/functionbeat/magefile.go
+++ b/x-pack/functionbeat/magefile.go
@@ -159,9 +159,11 @@ func BuildPkgForFunctions() error {
 	err := os.RemoveAll("pkg")
 
 	filesToCopy := map[string]string{
-		filepath.Join("provider", "aws", "functionbeat-aws"):      filepath.Join("pkg", "functionbeat-aws"),
-		filepath.Join("provider", "gcp", "pubsub", "pubsub.go"):   filepath.Join("pkg", "pubsub", "pubsub.go"),
-		filepath.Join("provider", "gcp", "storage", "storage.go"): filepath.Join("pkg", "storage", "storage.go"),
+		filepath.Join("provider", "aws", "functionbeat-aws"):           filepath.Join("pkg", "functionbeat-aws"),
+		filepath.Join("provider", "gcp", "pubsub", "pubsub.go"):        filepath.Join("pkg", "pubsub", "pubsub.go"),
+		filepath.Join("provider", "gcp", "storage", "storage.go"):      filepath.Join("pkg", "storage", "storage.go"),
+		filepath.Join("provider", "gcp", "build", "pubsub", "vendor"):  filepath.Join("pkg", "pubsub", "vendor"),
+		filepath.Join("provider", "gcp", "build", "storage", "vendor"): filepath.Join("pkg", "storage", "vendor"),
 	}
 	for src, dest := range filesToCopy {
 		c := &devtools.CopyTask{

--- a/x-pack/functionbeat/manager/gcp/template_builder.go
+++ b/x-pack/functionbeat/manager/gcp/template_builder.go
@@ -190,5 +190,6 @@ func zipResources() map[string][]bundle.Resource {
 
 func zipResourcesOfFunc(typeName string) []bundle.Resource {
 	root := filepath.Join("pkg", typeName)
-	return []bundle.Resource{&bundle.LocalFile{Path: filepath.Join(root, typeName+".go"), FileMode: 0755}}
+	vendor := bundle.Folder(filepath.Join("pkg", typeName, "vendor"), filepath.Join("pkg", typeName), 0644)
+	return append(vendor, &bundle.LocalFile{Path: filepath.Join(root, typeName+".go"), FileMode: 0755})
 }

--- a/x-pack/functionbeat/scripts/mage/update.go
+++ b/x-pack/functionbeat/scripts/mage/update.go
@@ -5,9 +5,13 @@
 package mage
 
 import (
+	"os"
+	"path/filepath"
+
 	"github.com/magefile/mage/mg"
 
 	devtools "github.com/elastic/beats/v7/dev-tools/mage"
+	"github.com/elastic/beats/v7/dev-tools/mage/gotool"
 )
 
 // Update target namespace.
@@ -20,7 +24,7 @@ var Aliases = map[string]interface{}{
 
 // All updates all generated content.
 func (Update) All() {
-	mg.Deps(Update.Fields, Update.IncludeFields, Update.Config, Update.FieldDocs)
+	mg.Deps(Update.Fields, Update.IncludeFields, Update.Config, Update.FieldDocs, Update.VendorBeats)
 }
 
 // Config generates both the short and reference configs.
@@ -45,4 +49,40 @@ func (Update) IncludeFields() error {
 	mg.Deps(Update.Fields)
 
 	return devtools.GenerateAllInOneFieldsGo()
+}
+
+// VendorBeats collects the vendor folder required to deploy the function for GCP.
+func (Update) VendorBeats() error {
+	for _, f := range []string{"pubsub", "storage"} {
+		gcpVendorPath := filepath.Join("provider", "gcp", "build", f, "vendor")
+		err := os.RemoveAll(gcpVendorPath)
+		if err != nil {
+			return err
+		}
+
+		deps, err := gotool.ListDepsLocation("github.com/elastic/beats/v7/x-pack/functionbeat/provider/gcp/" + f)
+		if err != nil {
+			return err
+		}
+
+		for importPath, location := range deps {
+			cp := &devtools.CopyTask{
+				Source:  location,
+				Dest:    filepath.Join(gcpVendorPath, importPath),
+				Mode:    0600,
+				DirMode: os.ModeDir | 0750,
+				Exclude: []string{
+					".*_test.go$",
+					".*.yml",
+				},
+			}
+			err = cp.Execute()
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	return nil
 }

